### ISSUE TITLE
Frontier support for Shoal client

### ISSUE THREAD

**Issue by rptaylor** [hep-gc/shoal#74]
Frontier server and proxy configuration is set by an environment variable:

```
[ ~]$  echo $FRONTIER_SERVER
(serverurl=http://frontier.triumf.ca:3128/ATLAS_frontier)(proxyurl=http://chrysaor.westgrid.ca:3128)(proxyurl=http://cernvm-webfs.atlas-canada.ca:3128)(serverurl=http://atlasfrontier-ai.cern.ch:8000/atlr)
```

We need to be able to invoke shoal-agent with some option in such a way that it outputs a string in the same format. The output string should be the same as the FRONTIER_SERVER environment variable of the parent process, with 1 or 2 additional phrases of the form  "(proxyurl=http://squid1:3128)"  where the squids are ones selected by Shoal.

**Comment by rptaylor** [comment]
So the usage might work something like this:

[rptaylor@symmetry01 ~]$ echo $FRONTIER_SERVER
(serverurl=http://frontier.triumf.ca:3128/ATLAS_frontier)(proxyurl=http://chrysaor.westgrid.ca:3128)(proxyurl=http://cernvm-webfs.atlas-canada.ca:3128)(serverurl=http://atlasfrontier-ai.cern.ch:8000/atlr)
[rptaylor@symmetry01 ~]$ outputstr=`shoal-client --frontier`
[rptaylor@symmetry01 ~]$ if [[ $? -eq 0 ]] ; then FRONTIER_SERVER=outputstr ; fi 
[rptaylor@symmetry01 ~]$ echo $FRONTIER_SERVER
(serverurl=http://frontier.triumf.ca:3128/ATLAS_frontier)(proxyurl=http://squid1.ca:3128)(proxyurl=http://squid2.ca:3128)(proxyurl=http://chrysaor.westgrid.ca:3128)(proxyurl=http://cernvm-webfs.atlas-canada.ca:3128)(serverurl=http://atlasfrontier-ai.cern.ch:8000/atlr)

**Comment by rptaylor** [comment]
When shoal-client is invoked in this mode, it might be useful to have a command-line option that specifies which Shoal server to use. If given, it overrides the one in shoal_client.conf. 

The reason for this is that we may have some info stored in and retrieved from AGIS for shoal+frontier usage.

This part could wait until Phase 2 of implementation.

**Comment by colsond** [comment]
The type of functionality in your last comment already exists within the client.
The shoal-client can be run with the --server option that allows you to specify a server at run-time.

Adding support for frontier shouldn't take long but before development starts I need to understand what you would expect shoal-client to output running with the --frontier option.

I see that in your output there is some serverurl's specified. Could you supply these to the client when you specify the frontier option or would shoal need to get them from somewhere else. I also notice that there is many squids specified in your output, would this be the expected functionality for frontier, and if so how many squids would you want added as default.

**Comment by rptaylor** [comment]
The shoal-client would need to read the original FRONTIER_SERVER environment variable as input.
Then the output (the serverurl and proxyurl strings) should be the same as the input, except with additional proxyurl strings inserted (2 of them by default) from Shoal.

**Comment by colsond** [comment]
@rptaylor  was support ever delivered for this // is this is still an enhancement that is desired?

**Comment by rptaylor** [comment]
Yes, it would still be useful to have ... not sure what happened to that commit from last year?

**Comment by colsond** [comment]
It seems like it was sort of neglected after I left. It looks like there has been some substantial changes to the client since then. I'll look at propagating those changes into the newest version. However, I may need your help to test it in the environment you plan on using it in.

**Comment by colsond** [comment]
I've updated the current code to facilitate frontier usage. I did a little testing on my own:
[~61 shoal-client]# export FRONTIER_SERVER="(serverurl=http://fakePreset123.ca:3128)"
[~61 shoal-client]# ./shoal-client -f
(serverurl=http://fakePreset123.ca:3128)(proxyurl=http://kraken01.westgrid.ca:3128)(proxyurl=http://atlascaq3.triumf.ca:3128)

I think these are the results you are looking for @rptaylor  but they should probably be tested in the environment you plan on using them in. 

I'm not sure if we are still building RPMS or just using ansible to deploy the components but I'll ask @kcasteels when he is in.

**Comment by rptaylor** [comment]
@colsond Yes, that is how it should work; that is consistent with my earlier post.

**Comment by MarcusEbert** [comment]
seems to do as expected, closing

**Comment by DrDaveD** [comment]
Consider also using the proxyconfigurl feature of frontier to process proxy auto config files.  ATLAS has done some work to support it but they hadn't quite finished it.  I think it will be finished soon, driven by a need for it with ATLAS@Home.
